File: server/src/notes_server/key_pool.py
import time
from dataclasses import dataclass
from typing import Literal

Status = Literal["ok", "cooling_down", "dead"]


@dataclass
class KeyState:
    key: str
    status: Status = "ok"
    cooldown_until: float = 0.0
    consecutive_errors: int = 0


class KeyPool:
    def __init__(self, keys: list[str], cooldown_seconds: int = 60):
        self._states: list[KeyState] = [KeyState(key=k) for k in keys]
        self._cooldown = cooldown_seconds
        self._cursor = 0

    def _is_available(self, s: KeyState) -> bool:
        if s.status == "dead":
            return False
        if s.status == "cooling_down" and time.time() < s.cooldown_until:
            return False
        if s.status == "cooling_down" and time.time() >= s.cooldown_until:
            s.status = "ok"
        return True

    def pick(self) -> str | None:
        n = len(self._states)
        for _ in range(n):
            s = self._states[self._cursor % n]
            self._cursor += 1
            if self._is_available(s):
                return s.key
        return None

    def _find(self, key: str) -> KeyState | None:
        return next((s for s in self._states if s.key == key), None)

    def mark_rate_limited(self, key: str, retry_after: float | None = None):
        s = self._find(key)
        if not s:
            return
        s.status = "cooling_down"
        s.cooldown_until = time.time() + (retry_after or self._cooldown)

    def mark_error(self, key: str):
        s = self._find(key)
        if not s:
            return
        s.consecutive_errors += 1
        if s.consecutive_errors >= 3:
            s.status = "dead"

    def mark_success(self, key: str):
        s = self._find(key)
        if not s:
            return
        s.consecutive_errors = 0
        if s.status != "dead":
            s.status = "ok"

    def summary(self) -> dict[str, dict]:
        return {
            s.key: {
                "status": s.status,
                "cooldown_until": s.cooldown_until,
                "consecutive_errors": s.consecutive_errors,
            }
            for s in self._states
        }

File: ios/NotesApp/AI/KeyPool.swift
import Foundation

/// Mirrors the server-side pool but runs on-device for school/fallback mode.
final class KeyPool {
    struct KeyState: CustomStringConvertible {
        var key: String
        var errorCount: Int = 0
        var cooldownUntil: Date? = nil
        var dead: Bool = false

        var description: String {
            let redacted = key.count > 6 ? "\(key.prefix(5))…\(key.suffix(2))" : "***"
            return "KeyState(\(redacted), errors=\(errorCount), dead=\(dead))"
        }
    }

    private var states: [KeyState]
    private var cursor: Int = 0
    private let clock: () -> Date
    private let lock = NSLock()

    init(keys: [String], clock: @escaping () -> Date = Date.init) {
        self.states = keys.map { KeyState(key: $0) }
        self.clock = clock
    }

    func pick() -> KeyState? {
        lock.lock(); defer { lock.unlock() }
        guard !states.isEmpty else { return nil }
        let now = clock()
        for _ in 0..<states.count {
            let idx = cursor % states.count
            cursor = (cursor + 1) % states.count
            var s = states[idx]
            if s.dead { continue }
            if let until = s.cooldownUntil, until > now { continue }
            if s.cooldownUntil != nil { s.cooldownUntil = nil }
            states[idx] = s
            return s
        }
        return nil
    }

    func markRateLimited(key: String, retryAfter seconds: TimeInterval) {
        lock.lock(); defer { lock.unlock() }
        guard let idx = states.firstIndex(where: { $0.key == key }) else { return }
        states[idx].cooldownUntil = clock().addingTimeInterval(seconds)
    }

    func markError(key: String) {
        lock.lock(); defer { lock.unlock() }
        guard let idx = states.firstIndex(where: { $0.key == key }) else { return }
        states[idx].errorCount += 1
        if states[idx].errorCount >= 3 {
            states[idx].dead = true
        }
    }

    func markSuccess(key: String) {
        lock.lock(); defer { lock.unlock() }
        guard let idx = states.firstIndex(where: { $0.key == key }) else { return }
        states[idx].errorCount = 0
    }

    var summary: [KeyState] {
        lock.lock(); defer { lock.unlock() }
        return states
    }
}
